File: src/linkedlist/Point arbit pointer to greatest value right side node in a linked list.c
#include<stdio.h>
#include<stdlib.h>
struct Node
{
    int data;
    struct Node* next, *arbit;
};

void add(int x,struct Node **head)
{
	struct Node *newnode=(struct Node*)malloc(sizeof(struct Node*)),*start;
	newnode->data=x;
	newnode->next=NULL;
	newnode->arbit=NULL;
	if(*head==NULL)
	{
		*head=newnode;
		return;
	}
	start=*head;
	while(start->next!=NULL)
		start=start->next;
	start->next=newnode;
}
void printNextArbitPointers(struct Node *node)
{
    printf("Node\tNext Pointer\tArbit Pointer\n");
    while (node!=NULL)
    {
        printf("%d\t\t",node->data);
 
        if (node->next)
            printf("%d\t\t",node->next->data);
        else printf("NULL\t\t");
 
        if (node->arbit)
            printf("%d\t\t\n",node->arbit->data);
        else printf("NULL\n");
 
        //printf("%d",endl);
        node = node->next;
    }
}

struct Node* nextGreater(struct Node *current)
{
	if(current==NULL||current->next==NULL)
		return current;
	struct Node* largest=nextGreater(current->next);
	current->arbit=largest;
	if(largest!=NULL&&current->data<largest->data)
		return largest;
	return current;
}
struct Node* populateArbit(struct Node *head)
{
	
	nextGreater(head);
	return head;
}
int main()
{
    struct  Node *head = NULL;
    add(5,&head);
	add(10,&head);
	add(20,&head);
	add(3,&head);
	head = populateArbit(head);
 
    printf("Resultant Linked List is: \n");
    printNextArbitPointers(head);
 
    return 0;
}

File: src/linkedlist/PointArbitPointerToGreatestValueRightSideNodeInALinkedList.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package linkedlist;

import java.util.LinkedList;

/**
 *
 * @author ashutosh
 */
public class PointArbitPointerToGreatestValueRightSideNodeInALinkedList
{
    public static void printWithArbit(Arbit node)
    {
        if(node==null)
            return;
        System.out.print(node.data+"\t\t");
        if(node.next==null)
            System.out.print("NULL\t\t");
        else System.out.print(node.next.data+"\t\t");
        if(node.arbit==null)
            System.out.print("NULL\t\t");
        else System.out.println(node.arbit.data);
        printWithArbit(node.next);
    }
    public static Arbit pointToArbit(Arbit node)
    {
        if(node==null||node.next==null)
            return node;
        Arbit largest=pointToArbit(node.next);
        node.arbit=largest;
        if(largest!=null&&node.data<largest.data)
            return largest;
        return node;
    }
    public static void main(String[] args)
    {
        LinkedList<Arbit> list=new LinkedList();
        Arbit five=new Arbit(1);
        Arbit ten=new Arbit(2);
        Arbit two=new Arbit(3);
        Arbit three=new Arbit(4);
        five.next=ten;ten.next=two;two.next=three;
        pointToArbit(five);
        System.out.println("Node\t\tNext\t\tArbit");
        printWithArbit(five);
    }
}
class Arbit
{
    Arbit arbit,next;
    int data;
    public Arbit(int data)
    {
        this.data=data;
        next=null;
        arbit=null;
    }
}